### DISCUSSION TITLE
how can I create multiple channel and re-establish endpoints?

### DISCUSSION THREAD

**Discussion by Fei-Ho** [OpenAMP/open-amp#574] [Q&A]
Can open-amp allow multiple clients and servers to communicate simultaneously through rpc?

The specific function description is as follows: Multiple pairs of clients and servers are started on the local host (linux) and remote processor (linux), and they communicate one-to-one based on rpc.

I tried it on the Linux host. I started two servers and clients repeatedly, but only the last pair started could execute normally.

**Comment by Fei-Ho** [comment]
I want to implement several one-to-one rpc communication processes between two cores. I tried using different value of name_service but failed, only one pair of process can work. Does open-amp support this? If possible, do you have any suggestions?

**Comment by arnopo** [comment]
Hello @Fei-Ho,
your set up is not clear, please give more details:
- do you use Linux rpmsg driver or OpenAMP on user space?
- do you implement your own rpc service or do you use [openamp rpmsg rpc service](https://github.com/OpenAMP/open-amp/tree/main/lib/service/rpmsg/rpc)?
- do you use rpmsg_ns_bind_cb ?

**Reply by Fei-Ho** [reply]
Thank you very much for your reply !

1. I use OpenAMP on user space
2. I use the [openamp rpmsg rpc service](https://github.com/OpenAMP/open-amp/tree/main/lib/service/rpmsg/rpc)
3. I'm not using rpmsg_ns_bind_cb

More detailed description:
I'm using the examples in open-amp/apps/examples/linux_rpc_demo/. But I simplified it: the client just calls a function on the server side that returns a fixed value. I now just try run server-side and client-side programs on my linux pc (ubuntu 20.04). Later, the server side will run on the device plugged into the linux host, and the client side will run on the linux host. What I want to achieve is illustrated in Figure 1 below:
![multi_channel](https://github.com/OpenAMP/open-amp/assets/30924686/05778a1d-f603-47db-bb90-63d771c39e3c)
Multiple servers need to be started on the remote processor side (core B), and they wait for rpc calls from the client on the local processor side (core A). I found that the ns parameter can only achieve matching between endpoints, that is, it can only achieve the method in Figure 2 below:
![one_channel](https://github.com/OpenAMP/open-amp/assets/30924686/fccb659e-e949-4611-8d00-b86a3cecc347)
How can I achieve the method in Figure 1?

In addition, I also encountered a problem: the server side can only handle one process call from the client side. When the client process finished and the client process is started again, the server does not respond. Similar to the issue described in this issue: https://github.com/OpenAMP/open-amp/issues/237. This issue says that re-establish endpoints or dynamic rediscovery of endpoints need to be implemented. Can you describe more specifically how to implement it?

Thank you so much !

**Comment by arnopo** [comment]
> Thank you very much for your reply !
> 
> 1. I use OpenAMP on user space
> 2. I use the [openamp rpmsg rpc service](https://github.com/OpenAMP/open-amp/tree/main/lib/service/rpmsg/rpc)
> 3. I'm not using rpmsg_ns_bind_cb
> 
> More detailed description: I'm using the examples in open-amp/apps/examples/linux_rpc_demo/. But I simplified it: the client just calls a function on the server side that returns a fixed value. I now just try run server-side and client-side programs on my linux pc (ubuntu 20.04). Later, the server side will run on the device plugged into the linux host, and the client side will run on the linux host. What I want to achieve is illustrated in Figure 1 below: ![multi_channel](https://private-user-images.githubusercontent.com/30924686/315869497-05778a1d-f603-47db-bb90-63d771c39e3c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEwOTQ2OTQsIm5iZiI6MTcxMTA5NDM5NCwicGF0aCI6Ii8zMDkyNDY4Ni8zMTU4Njk0OTctMDU3NzhhMWQtZjYwMy00N2RiLWJiOTAtNjNkNzcxYzM5ZTNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIyVDA3NTk1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViNTg5NDMyMzlhNTZmOTJiZDZjYTQ5OTI4YWJjMTEzM2MwOGRiYTI0NDI1NGU1M2I0YjE5YmZlMTljYTkzNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Mv0N9YEjHZDOE5foeMESHH0e5ivQoZoqOWCLW1ixACs) Multiple servers need to be started on the remote processor side (core B), and they wait for rpc calls from the client on the local processor side (core A). I found that the ns parameter can only achieve matching between endpoints, that is, it can only achieve the method in Figure 2 below: ![one_channel](https://private-user-images.githubusercontent.com/30924686/315869643-fccb659e-e949-4611-8d00-b86a3cecc347.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEwOTQ2OTQsIm5iZiI6MTcxMTA5NDM5NCwicGF0aCI6Ii8zMDkyNDY4Ni8zMTU4Njk2NDMtZmNjYjY1OWUtZTk0OS00NjExLThkMDAtYjg2YTNjZWNjMzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIyVDA3NTk1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1NjA1MWI1MDVhMjRkYjFiYTE4NjMxNGYzZTcxOTVmOWY0ODZjYWEwMmY5MjBhNjE1Y2ZiOTU2NzIwYWMxYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GgslVonz1dspHmeNW_dLPZZIV5RZcWMy6viG5ZLh_hY) How can I achieve the method in Figure 1?

Seems that the simpler way is to use name service to differentiate your servers,. you  could try to create a new API

```
int rpmsg_rpc_server_init_by_name(struct rpmsg_rpc_svr *rpcs, struct rpmsg_device *rdev,
			  const struct rpmsg_rpc_services *services, int len,
			  rpmsg_ns_unbind_cb rpmsg_service_server_unbind, const char *name)
{
	int ret;

	rpcs->services = services;
	rpcs->n_services = len;

	ret = rpmsg_create_ept(&rpcs->ept, rdev, name,
			       RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
			       rpmsg_endpoint_server_cb,
			       rpmsg_service_server_unbind);
	if (ret)
		return ret;

	return ret;
}
```
Same on the client side.


> 
> In addition, I also encountered a problem: the server side can only handle one process call from the client side. When the client process finished and the client process is started again, the server does not respond. Similar to the issue described in this issue: #237. This issue says that re-establish endpoints or dynamic rediscovery of endpoints need to be implemented. Can you describe more specifically how to implement it?
With the `rpmsg_service_server_unbind()` the server should be informed when the client endpoint is closed when `rpmsg_rpc_client_release() `is called. you could in this case restart a new server. 
 

> 
> Thank you so much !

**Reply by Fei-Ho** [reply]
Thank you very much for your reply !

I have tried this method and compiled two server executable files: server_1 and server_2, whose name services are set to rpmsg-rpc and rpmsg-rpc-2 respectively. client is also compiled to obtain two executable files: client_1 and client_2.
Start in the order of server_1, server_2, client_1, client_2. Both servers are unresponsive and are in a hang state.
Start in the order of server_1, server_2, client_2, client_1. Server_2 can complete the response to client_2, but server_1 does not respond and is in a hang state.

In addition, a bigger problem for me now is: when I keep the server process from ending and start the same client repeatedly, only the first time is valid, and subsequent starts will be in a hang state. I tried removing the code that the client sends ns destroy at the end but that didn't work. It seems that the client processes started later cannot find the server side.

**Reply by arnopo** [reply]
hang can come from a lot of different root cause.
First give a different name for the server string comparaison can fin a match between  rpmsg-rpc and rpmsg-rpc-2 using `strncmp`

Then if not solved  you can add some logs  to  print rpmsg_endpoint structure field sto verify the addr and dest_addr fields are valides on server and client side.

as test you can also update rpc server and client to force name to null and provide fix addresses ( instead of `RPMSG_ADDR_ANY`) to force the `addr `and `des_addr` of the endpoints
```
int rpmsg_rpc_server_init_by_addr(struct rpmsg_rpc_svr *rpcs, struct rpmsg_device *rdev,
			  const struct rpmsg_rpc_services *services, int len,
			  rpmsg_ns_unbind_cb rpmsg_service_server_unbind, uint32_t addr, uint32_t dest_addr)
{
	int ret;

	rpcs->services = services;
	rpcs->n_services = len;

	ret = rpmsg_create_ept(&rpcs->ept, rdev, NULL,
			       addr,  dest_addr,
			       rpmsg_endpoint_server_cb,
			       rpmsg_service_server_unbind);
	if (ret)
		return ret;

	return ret;
}
```

**Reply by Fei-Ho** [reply]
I found that the reasons why it is impossible to keep the server running and the client repeatedly starts to call the server is as follows: In the sk_unix_server() function in open-amp/apps/system/linux/generic/platform_info.c, the server-side is immediately closed its fd after the socket is created, resulting in the inability to continue to monitor other socket connection requests.

I briefly tested and found that commenting out "close(fd)" can achieve it, but a complete fix requires modifying many other things.

**Reply by Fei-Ho** [reply]
To clarify, the repeated calling of the client here means: starting the client process again after the execution of the client process is completed, rather than sending requests multiple times in one process.
